### PR TITLE
Fixed the messuring issue

### PR DESCRIPTION
**Pull request by georgi-nikolov**
There was a problem with when the textLayouts were created. The way the library was working resulted in the getText() returning null most of the time and hence the math that deals with how big the views should be was getting corrupted.

### CODE REVIEW

**Comment by xiphirx** [comment]
Can you revert the style changes and split the gradle changes into a separate PR?

**Comment by georgi-nikolov** [comment]
I have not made any style changes. All I did was move some code into the
onMeassure and onSizeChanged methods. I will revert the gradle changes in a
moment.

**Comment by xiphirx** [comment]
Your editor seems to have messed with the tab width. Please revert those changes.

**Comment by georgi-nikolov** [comment]
All done. Please check the code now. :)

2017-08-07 10:16 GMT+03:00 Hilal Alsibai <notifications@github.com>:

File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java [hunk 1]
```diff
@@ -614,7 +616,6 @@ void onDismiss(boolean userInitiated) {
       animator.removeAllUpdateListeners();
     }
 
-    ViewUtil.removeOnGlobalLayoutListener(getViewTreeObserver(), globalLayoutListener);
```
Comment (xiphirx) [inline taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java]: Do not remove this.

File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java [hunk 2]
```diff
@@ -348,7 +348,6 @@ public void onEnd() {
   private ValueAnimator[] animators = new ValueAnimator[]
       {expandAnimation, pulseAnimation, dismissConfirmAnimation, dismissAnimation};
 
-  private final ViewTreeObserver.OnGlobalLayoutListener globalLayoutListener;
```
Comment (xiphirx) [inline taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java]: ?

File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java [hunk 3]
```diff
@@ -510,6 +466,52 @@ public boolean onLongClick(View v) {
...
+  @Override
+  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
+    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
+    updateTextLayouts();
+  }
+
+  @Override
+  protected void onSizeChanged(int w, int h, int oldw, int oldh) {
+    super.onSizeChanged(w, h, oldw, oldh);
+    target.onReady(new Runnable() {
```
Comment (xiphirx) [inline taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java]: I'm not sure why this change is needed.

File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java [hunk 4]
```diff
@@ -840,7 +841,7 @@ void drawTintedTarget() {
   }
 
   void updateTextLayouts() {
-    final int textWidth = Math.min(getWidth(), TEXT_MAX_WIDTH) - TEXT_PADDING * 2;
+    final int textWidth = Math.min(getMeasuredWidth(), TEXT_MAX_WIDTH) - TEXT_PADDING * 2;
```
Comment (xiphirx) [inline taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java]: The measured width is not the correct value to use here.

File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java (hunk 3 above)
Comment (georgi-neykov-hub) [inline taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTargetView.java]: How do you handle the case where the `TapTargetView` instance gets re-measured on a layout pass. You add a `ViewTreeObserver` in the constructor and at some point you remove it, just to get the `View`'s size after a layout, which is somewaht strange given the fact that you can get what you need from various `On...` callbacks in the view.

The whole point of this change is to fix a bug where after a screen power off/on the view's circle is incorrectly sized and text width is 0. It can on any other event that will trigger a re-layout.

It might also solve the problem in [issue #91 ](https://github.com/KeepSafe/TapTargetView/issues/91).

**Comment by georgi-nikolov** [comment]
Hello again. I'd like to ask if you'd like to accept my pull request or at least discuss it further since we'd like to use your library and improve it if possible. It is easier for us not to fork the library just to fix a meassuring issue and I believe it is benefitial to you as well to have this issue fixed. If you'd like me to I can provide screenshots without my fix so you can see how the UI breaks. 
Best regards,
A fellow android dev :)

**Comment by xiphirx** [comment]
I need some code that shows the issue you're fixing. I'm honestly not entirely sure what this is fixing and it is making pretty large changes to how things function.

**Comment by georgi-nikolov** [comment]
Here is an example video to show how the code breaks. This is using the non forked version of your library - https://my.pcloud.com/publink/show?code=XZra8oZ9zbLYJbFJXjPK4T4jKhfU4R7u6ry 
Here is a snippet of how I am doing it in the code - https://pastebin.com/ufdJjiTA
Briefly explained i wait for a callback on the button of the snackbar, after that i programatically open my navigation drawer and wait for a callback that it is open and finally I find the exact view in the drawer i'd like to highlight and I generate the overlay with the provided builder in the library. Nothing fancy.
On the video whenever you see the picture freeze for a split second that is where I am turning the screen on and off via the lock button on the device. The bug occurs also in some cases with screen rotation - it either can't get the position of the view properly like in the video or it can't caclulate the size of the outter circle and the position of the text properly. Unfortuantelly I am unable to reproduce the later case with my current build. Both cases happen for the same reason - nowhere in the library do you honor the possibility that the system will call the layout again after the global layout observer has been removed and it leads to wrong math and broken UI in some cases like in the video. The proper solution in my oppinion is to move the logic for the size/position calculations of the overlay into a callback like onSizeChanged so whenever a new layout is calculated the overlay can recalculate its proper size and position as needed. The bug shown in the video is not reproducible any more with the fix i proposed in the pull request nor is the case where the size of the outher circle and/or text position is incorrect. Basically I believe both cases of the broken UI come down to not honoring the possibility of the system generating a layout after the tree observer is removed.

**Comment by xiphirx** [comment]
In your code, is there a reason why you're caching `tapTargetView` and not recreating it each time?

**Comment by georgi-nikolov** [comment]
I am recreating it on each new instance of the fragment I am using, however I am keeping a reference so i can clean up when I am not needing the tutorial to be showing anymore - by cleanUp i mean remove the view from the activity decor. Sometimes I need to cleanUp the tutorial without destroying the fragment right away and this is why i keep the reference. Each time I remove the tapTargetView from the activity decor i also null the reference so I am never actually reusing the same instance of tapTargetView

**Comment by xiphirx** [comment]
If the goal is to continue listening to layout changes from the targeted `View` just in case it changes, then what you have here won't work in all cases since `onSizeChanged` will only be called if the `TapTargetView`'s size changes, rather than the target's size, which wont be true past its first layout. Additionally, it wont account for translations and other transformations.

It seems like you want to use `View.onLayoutChange` instead. This is actually something I was planning to implement in v2 of this library, it would be cool to see it land in v1 if you're up for it. It should require minimal changes, but `TapTargetView` must ensure it cleans up it's `OnLayoutChangeListener` internally.

**Comment by georgi-nikolov** [comment]
I will take a look at that when I go to work in an hour or so and I will
write back. At first glance there is logic in what you're saying

**Comment by georgi-nikolov** [comment]
Hello,
I implemented things the way you suggested - https://pastebin.com/6H8DanH9
I kept a reference to a View.OnLayoutChangeListener and I put the code in the onLayoutChange callback. I also made sure to clean the listener in the onDismiss method. Unfortunatelly this does not fix the issue I showed in the video and I am able to reproduce it every time. I tried to reproduce the issue with my implementation (onSizeChanged) as well but I couldn't break the code then. Am I implementing your idea wrongly?

**Comment by xiphirx** [comment]
You want to have the layout listener attached to the target view. Currently there is no API to retrieve the underlying targeted view from the given `TapTarget` so youll have to add something like `getTargetedView` to `TapTarget` and attach the layout listener to that. The code you tried is adding a layout listener to `TapTargetView` itself, so its not paying attention to changes in the target view.

**Comment by xiphirx** [comment]
I'm going to close this out since there hasnt been any updates.
